numpy/lib/arraypad.py

```diff
@@ -885,105 +885,52 @@ def _pad_wrap(arr, pad_amt, axis=-1):
-        ((before_1, after_1), ... (before_N, after_N)) unique number of
-        elements for each axis where `N` is rank of `narray`.
-        ((before, after),) yields same before and after constants for each
-        axis.
-        (constant,) or val is a shortcut for before = after = constant for
-        all axes.
-    cast_to_int : bool, optional
-        Controls if values in ``shape`` will be rounded and cast to int
-        before being returned.
+    x : {None, scalar, array-like}
+        The object to broadcast to the shape (`ndim`, 2). None is only allowed
+        as a special case if `as_index` is True.
+    ndim : int
+        Number of pairs the broadcasted `x` will have.
+    as_index : bool, optional
+        If `x` is not None, try to round each element of `x` to a non-negative
+        integer.
 
     Returns
     -------
-    normalized_shape : tuple of tuples
-        val                               => ((val, val), (val, val), ...)
-        [[val1, val2], [val3, val4], ...] => ((val1, val2), (val3, val4), ...)
-        ((val1, val2), (val3, val4), ...) => no change
-        [[val1, val2], ]                  => ((val1, val2), (val1, val2), ...)
-        ((val1, val2), )                  => ((val1, val2), (val1, val2), ...)
-        [[val ,     ], ]                  => ((val, val), (val, val), ...)
-        ((val ,     ), )                  => ((val, val), (val, val), ...)
-
+    pairs : ndarray, shape (`ndim`, 2)
+        The broadcasted version of `x`.
+
+    Raises
+    ------
+    ValueError
+        If `as_index` is True and `x` contains negative elements.
+        Or if `x` is not broadcastable to the shape (`ndim`, 2).
     """
-    ndims = ndarray.ndim
-
-    # Shortcut shape=None
-    if shape is None:
-        return ((None, None), ) * ndims
+    if as_index and x is None:
+        # Pass through None as a special case for indices
+        return np.array(((None, None),) * ndim)
 
-    # Convert any input `info` to a NumPy array
-    shape_arr = np.asarray(shape)
+    x = np.asarray(x)
 
-    try:
-        shape_arr = np.broadcast_to(shape_arr, (ndims, 2))
-    except ValueError:
-        fmt = "Unable to create correctly shaped tuple from %s"
-        raise ValueError(fmt % (shape,))
-
-    # Cast if necessary
-    if cast_to_int is True:
-        shape_arr = np.round(shape_arr).astype(int)
-
-    # Convert list of lists to tuple of tuples
-    return tuple(tuple(axis) for axis in shape_arr.tolist())
-
-
-def _validate_lengths(narray, number_elements):
-    """
-    Private function which does some checks and reformats pad_width and
-    stat_length using _normalize_shape.
+    if x.ndim > 2:
+        # Fail explicitly when `x` already has more dimensions than
+        # desired shape (ndim, 2) accounts for
+        raise ValueError("keyword argument must not have more than 2 "
+                         "dimensions, got {}".format(x.ndim))
+    if as_index:
+        x = x.round().astype(np.intp, copy=False)
```

Comment:
Would be great to add a failing unit tests for this. Here's one that fails:
```
# pad the non-empty axis so that no memory allocation is needed
assert_equal(np.pad(np.zeros((1, 0)), [(0, 2**31), (0, 0)],  'constant').shape, (2**31+1, 0))
assert_equal(np.pad(np.zeros((1, 0)), [(0, 2**32), (0, 0)],  'constant').shape, (2**32+1, 0))
```

Reply:
Written up as #12023

Reply:
Will do. :)

Reply:
> Would be great to add a failing unit tests for this.

You mean marked with pytest's [xfail](https://docs.pytest.org/en/latest/skipping.html#xfail-mark-test-functions-as-expected-to-fail)? Considering the explanation in https://github.com/numpy/numpy/issues/12023#issuecomment-423921175:  how would I mark a unit test as xfailing only on Windows Systems?

Reply:
Wouldn't it be simpler to test `_as_pairs` directly if the unit test in question would be introduced alongside this anyway?

E.g.:

```python
_as_pairs([np.iinfo(np.intp).max], 2, as_index=True)   
# -> passes
_as_pairs([np.iinfo(np.intp).max + 1], 2, as_index=True)   
# -> raises "ValueError: index can't contain negative values"
```

I'm still not sure what this test is supposed to accomplish. You want to document the behavior for the edge case around the maximum of inpt, correct?

Also 
```python
np.pad(np.zeros((1, 0)), [(0, np.iinfo(np.intp).max), (0, 0)],  'constant')
```
will raise a different error which may not be what you intended...
```
ValueError: array is too big; `arr.size * arr.dtype.itemsize` is larger than the maximum possible size.
```

Reply:
@mattip, @mhvk Before you consider merging this I just want to highlight this thread in case you missed it. From what I gathered @eric-wieser suggested this as an optional improvement. However I still don't know how to proceed.

Of course I can always add the unit test as a separate PR if / when this thread is resolved.